### PR TITLE
Partially implements IMEManager.

### PR DESCRIPTION
**Pull request by tomsons26**

### CODE REVIEW

File: src/game/client/gui/imemanagerinterface.h [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+class Utf16String;
+
+class IMEManagerInterface : public SubsystemInterface
+{
+public:
+    virtual ~IMEManagerInterface() {}
+
+    virtual void Init() = 0;
+    virtual void Reset() = 0;
+    virtual void Update() = 0;
```
Comment (duncanspumpkin) [inline src/game/client/gui/imemanagerinterface.h]: these are just inherited and dont need to be here.

File: src/game/client/gui/imemanager.cpp [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+Utf16String *IMEManager::Get_Candidate(int index)
+{
+    if (m_candidateString && index >= 0 && index < m_candidateCount) {
+        return &m_candidateString[index];
+    }
+
+    // TODO this may be dangerous, need to find out if candidate is modified...
+    return (Utf16String *)&Utf16String::s_emptyString;
```
Comment (xezon) [inline src/game/client/gui/imemanager.cpp:132]: return `const Utf16String *`, remove the cast, then it is safe.

Also I suspect it returns reference, not pointer. Is unusual to return pointer this way if it is not expected to be null.

File: src/game/client/gui/imemanagerinterface.h [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    virtual int Get_Composition_Cursor_Position() = 0;
+
+    virtual int Get_Index_Base() = 0;
+
+    virtual int Get_Candidate_Count() = 0;
+    virtual Utf16String *Get_Candidate(int index) = 0;
+    virtual int Get_Selected_Candidate_Index() = 0;
+
+    virtual int Get_Candidate_Page_Size() = 0;
+    virtual int Get_Candidate_Page_Start() = 0;
```
Comment (xezon) [inline src/game/client/gui/imemanagerinterface.h]: Can we make these virtual get functions `const` ?

File: src/game/client/gui/imemanager.cpp (hunk 2 above)
Comment (tomsons26) [reply src/game/client/gui/imemanager.cpp:132]: i suspect candidate is modified

**Pull request merged**